GitHubBranches.ps1

```diff
@@ -154,6 +154,518 @@ filter Get-GitHubRepositoryBranch
+
+        [string[]] $StatusChecks,
+
+        [switch] $RequireUpToDateBranches,
+
+        [switch] $EnforceAdmins,
+
+        [string[]] $DismissalUsers,
+
+        [string[]] $DismissalTeams,
+
+        [switch] $DismissStaleReviews,
+
+        [switch] $RequireCodeOwnerReviews,
+
+        [ValidateRange(1,6)]
+        [int] $RequiredApprovingReviewCount,
+
+        [string[]] $RestrictPushUsers,
+
+        [string[]] $RestrictPushTeams,
+
+        [string[]] $RestrictPushApps,
+
+        [switch]$RequiredLinearHistory,
+
+        [switch] $AllowForcePushes,
+
+        [switch] $AllowDeletions,
+
+        [string] $AccessToken,
+
+        [switch] $NoStatus
+    )
+
+    $elements = Resolve-RepositoryElements
+    $OwnerName = $elements.ownerName
+    $RepositoryName = $elements.repositoryName
+
+    $telemetryProperties = @{
+        OwnerName = (Get-PiiSafeString -PlainText $OwnerName)
+        RepositoryName = (Get-PiiSafeString -PlainText $RepositoryName)
+    }
+
+    if ($PSBoundParameters.ContainsKey('StatusChecks'))
+    {
+        $requiredStatusChecks = @{
+            strict = $RequireUpToDateBranches.ToBool()
+            contexts = $StatusChecks
+        }
+    }
+    else
+    {
+        $requiredStatusChecks = $null
+    }
+
+    $dismissalRestrictions = @{}
+
+    if ($PSBoundParameters.ContainsKey('DismissalUsers')) {
+       $dismissalRestrictions['users'] = $DismissalUsers }
+    if ($PSBoundParameters.ContainsKey('DismissalTeams'))
+    {
+        $teams = Get-GitHubTeam -OwnerName $OwnerName -RepositoryName $RepositoryName |
+            Where-Object -FilterScript { $DismissalTeams -contains $_.name }
+        $dismissalRestrictions['teams'] = @($teams.slug)
+    }
+
+    $requiredPullRequestReviews = @{}
+
+    if ($PSBoundParameters.ContainsKey('DismissStaleReviews')) {
+        $requiredPullRequestReviews['dismiss_stale_reviews'] = $DismissStaleReviews.ToBool() }
+    if ($PSBoundParameters.ContainsKey('RequireCodeOwnerReviews')) {
+        $requiredPullRequestReviews['require_code_owner_reviews'] = $RequireCodeOwnerReviews.ToBool() }
+    if ($PSBoundParameters.ContainsKey('RequiredApprovingReviewCount')) {
+        $requiredPullRequestReviews['required_approving_review_count'] = $RequiredApprovingReviewCount }
+
+    if ($dismissalRestrictions.count -gt 0)
+    {
+        $requiredPullRequestReviews['dismissal_restrictions'] = $dismissalRestrictions
+    }
+
+    if ($requiredPullRequestReviews.count -eq 0)
+    {
+        $requiredPullRequestReviews = $null
+    }
+
+    if ($PSBoundParameters.ContainsKey('RestrictPushUsers') -or
+        $PSBoundParameters.ContainsKey('RestrictPushTeams') -or
+        $PSBoundParameters.ContainsKey('RestrictPushApps'))
+    {
+        if ($null -eq $RestrictPushUsers)
+        {
+            $RestrictPushUsers = @()
+        }
+
+        if ($null -eq $RestrictPushTeams)
+        {
+            $RestrictPushTeams = @()
+        }
+
+        $restrictions = @{
```

Comment:
Similar to my feedback above regarding `required_status_checks`, this will introduce regressions if users try to update _just_ the push users or _just_ the teams.  If I already have values defined for users, and now I go and want to make a separate call to set the teams, this code, as written, will wipe out my defined users.

Reply:
GitHub have really over-complicated this API and it doesn't allow you to do this. To achieve this functionality, all the other individual API endpoints would need to be implemented. I was originally thinking of making this function `New-GitHubRepositoryBranchProtectionRule` and am happy to go back to that if required.

Reply:
I don't follow.  You've already implemented `Get-GitHubRepositoryBranchProtectionRule`.  If you detected that any of these object properties were being altered and called that first, wouldn't you have all the content that you'd need in order to ensure you weren't overwriting any property that the user didn't provide in this call?